design_documents/accessing_emitted_events.md

```diff
@@ -0,0 +1,191 @@
+        fn emit_store_name(self: @ContractState, name: felt252) {
+            // ...
+            self.emit(Event::StoredName(StoredName { user: get_caller_address(), name: name }));
+        }
+    }
+}
+
+use snforge_std::expect_events;
+use snforge_std::Event;
+
+#[test]
+fn check_emitted_event() {
+    // ...
+	expect_events(array![Event { name: 'StoredName', keys: array![123], data: array![456] }]);
+    let res = contract.emit_store_name(...);  // if the event is not emitted it will panic
+
+    let res = contract.emit_store_name(...);  // expect_events does not work here
+    
+    expect_exact_events(array![Event { name: 'StoredName', keys: array![123], data: array![456] }]);
+    let res = contract.emit_store_name(...);  // function has to emit exactly those events defined in the array otherwise it will panic
+    // ...
+}
+```
+
+## `spy_events` cheatcode
+
+Another idea is to give users the highest possible customization. There would be a handler for extracting events emitted
+after the line it was created. Users would have to assert events manually (which could be more complex than the previous
+solution but would add more flexibility).
+
+Introduce `spy_events` cheatcode with the signature:
+
+```cario
+fn spy_events() -> snforge_std::EventSpy
+```
+
+where `snforge_std::EventSpy` would allow for accessing events emitted after its creation.
+`EventSpy` would be defined as follows.
+
+```cario
+struct EventSpy {
+    events: Array<snforge_std::Event>,
+}
+
+struct Event {
+    from: ContractAddress,
+    name: felt252,
+    keys: Array<felt252>,
+    data: Array<felt252>
+}
+
+trait EventFetcher {
+    fn fetch_events(ref self: EventSpy);
+}
+
+trait EventAssertions {
+    fn assert_emitted(ref self: EventSpy, events: Array<snforge_std::Event>);
+}
+```
+
+Users will be responsible for calling `fetch_events` to load emitted events to the `events` property.
+
+### Usage example
+
+```cario
+// Contract is the same as in the previous example
+
+use snforge_std::spy_events;
+use snforge_std::EventSpy;
+use snforge_std::EventFetcher;
+use snforge_std::EventAsserions;
+
+#[test]
+fn check_emitted_event() {
+    // ...
+	let mut spy = spy_events();  // all events emitted after this line will be saved under the `spy` variable
+    let res = contract.emit_store_name(...);
+    
+    spy.fetch_events();
+    
+    // users can assert events on their own
+    assert(spy.events.len() == 1, 'There should be one event');
+    assert(spy.events.at(0).name == sn_keccak('StoredName'), 'Wrong event name');
+    
+    let data = array![...];
+    assert(spy.events.at(0).data == data, 'Wrong data');
+    
+    // or use prepared assertions
+    spy.assert_emitted(array![Event {from: ..., name: 'StoredName', ...}, Event {...}]);
+
+    let res = contract.emit_store_name(...);
+    let res = contract.emit_store_name(...);
+    
+    // events will not be present before fetching
+    assert!(spy.events.len() == 1, 'There should be one event');
```

Comment:
I think the events array should be emptied every time `assert_emitted` is called. Otherwise `assert_emitted` loses its purpose when used multiple times, as we match against previous events as well

Reply:
What if someone would like to check if an event was emitted and then perform some custom asserts? I'm not sure if it's a valid case but theoretically could happen 🤷

Reply:
Well that depends what do we want to put emphasis on. Imo `assert_emitted` should provide a simple and intuitive way to check for events occurrences, so that user can easily insert it into test flow and bother neither about internal behaviours (like the one mentioned above) nor calling other functions (like `fetch_events`). If someone wants to perform custom asserts, they can use fetch_events and the `events` field